helpers/sql-obfuscation/lib/opentelemetry/helpers/sql_obfuscation.rb

```diff
@@ -0,0 +1,147 @@
+        comments: /(?:#|--).*?(?=\r|\n|$)/i,
+        multi_line_comments: %r{\/\*(?:[^\/]|\/[^*])*?(?:\*\/|\/\*.*)}
+      }.freeze
+
+      DIALECT_COMPONENTS = {
+        default: COMPONENTS_REGEX_MAP.keys,
+        mysql: %i[
+          single_quotes
+          double_quotes
+          numeric_literals
+          boolean_literals
+          hexadecimal_literals
+          comments
+          multi_line_comments
+        ],
+        postgres: %i[
+          single_quotes
+          dollar_quotes
+          uuids
+          numeric_literals
+          boolean_literals
+          comments
+          multi_line_comments
+        ]
+      }.freeze
+
+      PLACEHOLDER = '?'
+      UNMATCHED_PAIRS_FAILURE_MESSAGE = 'Failed to obfuscate SQL query - quote characters remained after obfuscation'
+      OBFUSCATION_FAILURE_MESSAGE = 'Failed to obfuscate SQL'
+
+      # We use these to check whether the query contains any quote characters
+      # after obfuscation. If so, that's a good indication that the original
+      # query was malformed, and so our obfuscation can't reliably find
+      # literals. In such a case, we'll replace the entire query with a
+      # placeholder.
+      CLEANUP_REGEX = {
+        default: %r{'|"|\/\*|\*\/},
+        mysql: %r{'|"|\/\*|\*\/},
+        postgres: %r{'|\/\*|\*\/|\$(?!\?)}
+      }.freeze
+
+      # @api private
+      def self.generate_regex(dialect)
+        components = DIALECT_COMPONENTS[dialect]
+        Regexp.union(components.map { |component| COMPONENTS_REGEX_MAP[component] })
+      end
+
+      DEFAULT_COMPONENTS_REGEX = generate_regex(:default)
+      MYSQL_COMPONENTS_REGEX = generate_regex(:mysql)
+      POSTGRES_COMPONENTS_REGEX = generate_regex(:postgres)
+
+      # This is a SQL obfuscation utility intended for use in database adapter instrumentation.
+      #
+      # @param sql [String] The SQL to obfuscate.
+      # @param obfuscation_limit [optional Integer] The maximum length of an obfuscated sql statement.
+      # @param adapter [optional Symbol] the type of database adapter calling the method. `:default`, `:mysql` and `:postgres` are supported.
+      # @return [String] The SQL query string where the values are replaced with "?". When the sql statement exceeds the obufscation limit
+      #  the first matched pair from the SQL statement will be returned, with an appended truncation message. If trunaction is unsuccessful,
+      #  a string describing the error will be returned.
+      #
+      # @api public
+      def self.obfuscate_sql(sql, obfuscation_limit: 2000, adapter: :default)
+        regex = case adapter
+                when :mysql
+                  MYSQL_COMPONENTS_REGEX
+                when :postgres
+                  POSTGRES_COMPONENTS_REGEX
+                else
+                  DEFAULT_COMPONENTS_REGEX
+                end
+
+        return truncate_statement(sql, regex, obfuscation_limit) if sql.size > obfuscation_limit
```

Comment:
Looking at this again. Is there a use case where the truncate statement fails because we have not encoded the SQL as utf8?

Reply:
Great question. I'll test this case and get back to you.

Reply:
After looking into it, the `truncate_statement` call wouldn't fail due to a lack of UTF-8 encoding. However, the errors the UTF-8 encoding method is trying to prevent could still occur downstream.

I updated the code to UTF-8 encode before truncation and then ran `gsub` after the obfuscation limit was checked. This changes the original functionality slightly, but it's a bit more robust to protect against UTF-8 errors.

See: 2805b6dc8f10012bce8afb794dc8e376ca30cc44

Reply:
Hmm... it looks like this change has a few issues with MySql2 and Trilogy tests. Since we're changing the behavior of the original obfuscation code here, would you be open to breaking this code into a bugfix and including it in a separate release? I kind of like the idea of keeping things as consistent as possible with the original obfuscation code for the first release.

Reply:
Ah, actually the error was introduced when I switched the `OpenTelemetry.logger.debug` call for the `OpenTelemetry.handle_error` call. Not quite sure what's going on here... I'll investigate.

Reply:
That is because we are handling the error so the last line of the method call is the standard error instead of a value you intended on returning. 

```ruby
def example(x)
  if x == 1
    x
  else
    raise "this is an error"
  end
  rescue StandardError => e
    "Unexpected #{e.message}"
end

puts example(1)
puts example(2)
```



```console
🤘 ruby return_last_line.rb
1
Unexpected this is an error
``` 

So in our case the method is returning the result of `handle_error` or `debug`

Reply:
Either way I think the result of this message should be the `OBFUSCATION_FAILURE_MESSAGE` when a StandardError occurs.